js/errors.ts

```diff
@@ -1,17 +1,83 @@
+// Copyright 2018 the Deno authors. All rights reserved. MIT license.
 import { deno as fbs } from "gen/msg_generated";
 
-// @internal
+const ERR_PREFIX = "Err";
+
 export class DenoError<T extends fbs.ErrorKind> extends Error {
   constructor(readonly kind: T, msg: string) {
     super(msg);
-    this.name = `deno.${fbs.ErrorKind[kind]}`;
+    this.name = `deno.${ERR_PREFIX}${fbs.ErrorKind[kind]}`;
   }
 }
 
+const errorClasses = new Map();
+
+function ErrorFactory<T extends fbs.ErrorKind>(
+  kind: T
+): new (msg: string) => DenoError<T> {
+  const name = `${ERR_PREFIX}${fbs.ErrorKind[kind]}`;
+  const anonymousClass = class extends DenoError<T> {
+    constructor(msg: string) {
+      super(kind, msg);
+    }
+  };
+  Object.defineProperty(anonymousClass, "name", {
+    value: name,
+  });
+  errorClasses.set(kind, anonymousClass);
+  return anonymousClass;
+}
+
 // @internal
 export function maybeThrowError(base: fbs.Base): void {
   const kind = base.errorKind();
   if (kind !== fbs.ErrorKind.NoError) {
-    throw new DenoError(kind, base.error()!);
+    const errorClass = errorClasses.get(kind);
+    throw new errorClass(base.error()!);
   }
 }
+
+export const errorNames: string[] = Object.keys(fbs.ErrorKind)
+  .filter(x => isNaN(Number(x)) && x !== "NoError").map(e => ERR_PREFIX + e);
```

Comment:
This seems a little redundant, since Number("NoError") is NaN. 
I wonder what other non-numbers you are expecting?

Reply:
@piscisaureus  It's a combination of two things:
1) `x` must not be a numeric string.
2) Remove `NoError` from `errorNames` because it's not an error.
FYI;
```ts
Object.keys(fbs.ErrorKind) == ["0", "NoError", "1", "NotFound", ...]
```